File: backend/app/schemas/application.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel

from app.domain.enums import AppStatus


class CreateApplicationRequest(BaseModel):
    program_id: uuid.UUID
    period_id: uuid.UUID


class ApplicationCreatedResponse(BaseModel):
    application_id: uuid.UUID
    status: str


class ApplicationSummary(BaseModel):
    id: uuid.UUID
    program_id: uuid.UUID
    period_id: uuid.UUID
    status: str
    tracking_number: Optional[str]
    submitted_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}


class AcademicRecordResponse(BaseModel):
    institution: Optional[str]
    gpa_4: Optional[Decimal]
    gpa_100: Optional[Decimal]
    yks_score: Optional[Decimal]
    credits_completed: Optional[int]
    fetched_at: Optional[datetime]
    source: Optional[str]
    errors: Optional[List[str]] = None

    model_config = {"from_attributes": True}


class EligibilityCheckResponse(BaseModel):
    rule_key: str
    passed: bool
    detail: Optional[str]

    model_config = {"from_attributes": True}


class ApplicationDetailResponse(BaseModel):
    id: uuid.UUID
    applicant_id: uuid.UUID
    program_id: uuid.UUID
    period_id: uuid.UUID
    status: str
    tracking_number: Optional[str]
    submitted_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    progress: dict
    eligibility_checks: List[EligibilityCheckResponse]

    model_config = {"from_attributes": True}


class SubmitApplicationResponse(BaseModel):
    tracking_number: str
    status: str


class StatusChangeRequest(BaseModel):
    new_status: AppStatus
    note: Optional[str] = None


# ---------------------------------------------------------------------------
# SPEC-005 — Status / History
# ---------------------------------------------------------------------------

class StageOut(BaseModel):
    name: str
    label_tr: str
    label_en: str
    completed: bool
    active: bool


class ProgressOut(BaseModel):
    stages: List[StageOut]
    current_stage: str


class HistoryEntry(BaseModel):
    status: str
    changed_at: datetime
    changed_by_role: Optional[str]
    note: Optional[str]


class ResultOut(BaseModel):
    outcome: str
    reason: Optional[str]


class ApplicationStatusResponse(BaseModel):
    tracking_number: Optional[str]
    status: str
    progress: ProgressOut
    history: List[HistoryEntry]
    result: Optional[ResultOut]

File: frontend/src/types/application.ts
export type AppStatus =
  | 'DRAFT'
  | 'SUBMITTED'
  | 'UNDER_REVIEW'
  | 'ENGLISH_REVIEW'
  | 'DEPT_EVAL'
  | 'RANKING'
  | 'DEAN_APPROVED'
  | 'ANNOUNCED'
  | 'REJECTED'
  | 'CORRECTION_REQUESTED'

export type DocType =
  | 'TRANSCRIPT'
  | 'YKS_RESULT'
  | 'LANGUAGE_CERT'
  | 'ID_COPY'
  | 'MILITARY_STATUS'
  | 'DISCIPLINE_RECORD'
  | 'OTHER'

export type DocStatus = 'PENDING' | 'ACCEPTED' | 'REJECTED' | 'CORRECTION_REQUESTED'

export interface ApplicationSummary {
  id: string
  program_id: string
  period_id: string
  status: AppStatus
  tracking_number: string | null
  submitted_at: string | null
  created_at: string
}

export interface ProgressStep {
  step: string
  completed: boolean
  active: boolean
  pending: boolean
}

export interface ApplicationDetail extends ApplicationSummary {
  applicant_id: string
  updated_at: string
  progress: {
    tracking_number: string | null
    current_status: AppStatus
    steps: ProgressStep[]
    percentage: number
    is_terminal: boolean
  }
  eligibility_checks: EligibilityCheck[]
}

export interface EligibilityCheck {
  rule_key: string
  passed: boolean
  detail: string | null
}

export interface AcademicRecord {
  institution: string | null
  gpa_4: number | null
  gpa_100: number | null
  yks_score: number | null
  credits_completed: number | null
  fetched_at: string | null
  source: string | null
  errors: string[] | null
}

export interface Document {
  id: string
  application_id: string
  doc_type: DocType
  file_name: string
  file_size_bytes: number | null
  status: DocStatus
  uploaded_at: string
  extracted_data: Record<string, unknown> | null
  extraction_confirmed: boolean
}

export interface PresignedUploadResult {
  upload_url: string
  object_key: string
}

export interface SubmitResult {
  tracking_number: string
  status: AppStatus
}

export interface PipelineStage {
  name: string
  label_tr: string
  label_en: string
  completed: boolean
  active: boolean
}

export interface HistoryEntry {
  status: AppStatus
  changed_at: string
  changed_by_role: string | null
  note: string | null
}

export interface ApplicationStatus {
  tracking_number: string | null
  status: AppStatus
  progress: {
    stages: PipelineStage[]
    current_stage: string
  }
  history: HistoryEntry[]
  result: { outcome: string; reason: string | null } | null
}

export interface NotificationMessage {
  id: string
  subject: string | null
  message: string
  status: 'PENDING' | 'SENT' | 'FAILED'
  sent_at: string | null
  created_at: string
}
